plugin/src/main/java/org/opensearch/ml/model/MLModelAndNodeManager.java

```diff
@@ -0,0 +1,294 @@
+import org.opensearch.ml.common.transport.load.LoadModelResponse;
+import org.opensearch.ml.common.transport.load.MLLoadModelAction;
+import org.opensearch.ml.common.transport.load.MLLoadModelRequest;
+import org.opensearch.ml.common.transport.model.MLModelSearchAction;
+import org.opensearch.ml.engine.MLEngine;
+import org.opensearch.ml.engine.ModelHelper;
+import org.opensearch.ml.indices.MLIndicesHandler;
+import org.opensearch.ml.stats.MLStats;
+import org.opensearch.ml.task.MLTaskManager;
+import org.opensearch.ml.utils.MLNodeUtils;
+import org.opensearch.rest.RestStatus;
+import org.opensearch.search.SearchHit;
+import org.opensearch.search.SearchHits;
+import org.opensearch.search.builder.SearchSourceBuilder;
+import org.opensearch.threadpool.ThreadPool;
+
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * Manager class for ML models and nodes. It contains ML model auto reload operations etc.
+ */
+@Log4j2
+public class MLModelAndNodeManager {
+    private final Client client;
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+    private final NamedXContentRegistry xContentRegistry;
+    private final ModelHelper modelHelper;
+    private final DiscoveryNodeHelper nodeHelper;
+    private final MLModelCacheHelper modelCacheHelper;
+    private final MLStats mlStats;
+    private final MLCircuitBreakerService mlCircuitBreakerService;
+    private final MLIndicesHandler mlIndicesHandler;
+    private final MLTaskManager mlTaskManager;
+    private final MLEngine mlEngine;
+    private volatile Boolean enableAutoReLoadModel;
+
+    public MLModelAndNodeManager(
+        ClusterService clusterService,
+        Client client,
+        ThreadPool threadPool,
+        NamedXContentRegistry xContentRegistry,
+        ModelHelper modelHelper,
+        DiscoveryNodeHelper nodeHelper,
+        Settings settings,
+        MLStats mlStats,
+        MLCircuitBreakerService mlCircuitBreakerService,
+        MLIndicesHandler mlIndicesHandler,
+        MLTaskManager mlTaskManager,
+        MLModelCacheHelper modelCacheHelper,
+        MLEngine mlEngine
+    ) {
+        this.clusterService = clusterService;
+        this.client = client;
+        this.threadPool = threadPool;
+        this.xContentRegistry = xContentRegistry;
+        this.modelHelper = modelHelper;
+        this.nodeHelper = nodeHelper;
+        this.modelCacheHelper = modelCacheHelper;
+        this.mlStats = mlStats;
+        this.mlCircuitBreakerService = mlCircuitBreakerService;
+        this.mlIndicesHandler = mlIndicesHandler;
+        this.mlTaskManager = mlTaskManager;
+        this.mlEngine = mlEngine;
+
+        enableAutoReLoadModel = ML_COMMONS_MODEL_AUTO_RELOAD_ENABLE.get(settings);
+        clusterService
+            .getClusterSettings()
+            .addSettingsUpdateConsumer(ML_COMMONS_MODEL_AUTO_RELOAD_ENABLE, it -> enableAutoReLoadModel = it);
+    }
+
```

Comment:
Using a new thread and pass necessary variables to this thread to execute autoReload action, after the retry logic, the thread should be destroyed.